src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,234 @@
+    return error;
+}
+
+CHIP_ERROR ServicePool::RemoveService(uint64_t nodeId, uint64_t fabricId)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+
+    if (nodeId == kUndefinedNodeId)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    do
+    {
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            break;
+        }
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == nodeId && mEntries[i].mFabricId == fabricId)
+        {
+            mEntries[i].mLazyDelete = true;
+            mLazyDeleteCount++;
+            FreeTextEntries(&mEntries[i].mService);
+        }
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ServicePool::UpdateService(uint64_t nodeId, uint64_t fabricId, const MdnsService & service)
+{
+    MdnsService * foundService;
+
+    if (!FindService(nodeId, fabricId, &foundService))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    FreeTextEntries(foundService);
+    *foundService = service;
+    return CopyTextEntries(foundService, service);
+}
+
+bool ServicePool::FindService(uint64_t nodeId, uint64_t fabricId, MdnsService ** service)
+{
+    size_t hashValue = nodeId % kServicePoolCapacity;
+    size_t i         = hashValue;
+
+    if (nodeId == kUndefinedNodeId)
+    {
+        *service = nullptr;
+        return false;
+    }
+
+    do
+    {
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == kUndefinedNodeId)
+        {
+            break;
+        }
+        if (!mEntries[i].mLazyDelete && mEntries[i].mNodeId == nodeId && mEntries[i].mFabricId == fabricId)
+        {
+            *service = &mEntries[i].mService;
+            return true;
+        }
+        i++;
+        i %= kServicePoolCapacity;
+    } while (i != hashValue);
+
+    *service = nullptr;
+    return false;
+}
+
+void ServicePool::Clear()
+{
+    for (size_t i = 0; i < kServicePoolCapacity; i++)
+    {
+        if (mEntries[i].mNodeId != kUndefinedNodeId)
+        {
+            FreeTextEntries(&mEntries[i].mService);
+        }
+        mEntries[i].mNodeId     = kUndefinedNodeId;
+        mEntries[i].mLazyDelete = false;
+    }
+    mLazyDeleteCount = 0;
+}
+
+void ServicePool::ClearNoFree()
+{
+    for (size_t i = 0; i < kServicePoolCapacity; i++)
+    {
+        mEntries[i].mNodeId     = kUndefinedNodeId;
+        mEntries[i].mLazyDelete = false;
+    }
+    mLazyDeleteCount = 0;
+}
+
+void ServicePool::ReHash()
+{
+    ServicePoolEntry copyEntries[kServicePoolCapacity];
+
+    memcpy(&copyEntries, mEntries, sizeof(mEntries));
+    ClearNoFree();
+
+    for (size_t i = 0; i < kServicePoolCapacity; i++)
+    {
+        if (!copyEntries[i].mLazyDelete && copyEntries[i].mNodeId != kUndefinedNodeId)
+        {
+            AddService(copyEntries[i].mNodeId, copyEntries[i].mFabricId, copyEntries[i].mService, /* deepCopy = */ false);
```

Comment:
It seems to me we should make the four-arg version of `AddService` private and have a public three-arg version; that way consumers won't accidentally call this sharp-edged one.